### DISCUSSION TITLE
Calling item functions on the result of a transform

### DISCUSSION THREAD

**Discussion by ltrzesniewski** [dotnet/msbuild#6687] [Q&A]
Apparently, the following works:

```XML
<ItemGroup>
  <Bar Include="@(Foo->'%(Bar).Baz'->ClearMetadata())" />
</ItemGroup>
```

I can call an item function (`ClearMetadata()`) on the result of a transform expression which looks like a string, and it works:

![image](https://user-images.githubusercontent.com/7913492/126156834-759363d5-0079-46c2-bdb8-0ef80ee4538f.png)

But string functions also work:

```XML
<ItemGroup>
  <Bar Include="@(Foo->'%(Bar).Baz'->Replace('Baz', 'FooBar'))" />
</ItemGroup>
```

![image](https://user-images.githubusercontent.com/7913492/126156880-bea28857-8530-4564-9d19-842ffc66b9ab.png)

I'd like to know if calling an item function like this is really supported, or if it just works by coincidence? The syntax looks unintuitive.

The [transforms documentation](https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-transforms) doesn't mention the type of the transform result, and a quick glance on the .NET SDK revealed only one place where an item function is used that way. I didn't expect this to work, since the transform looks like it produces a string, not an item.

Also, I didn't really expect item metadata to be preserved after a transform (since I thought the transform produces a string in the first place), so I suppose that clarifying this in the docs would be nice.

**Comment by KalleOlaviNiemitalo** [comment]
It doesn't look like a coincidence to me, but not really documented either, so I'm not sure it will work in future versions.

ExpressionShredder.GetReferencedItemExpressions parses the chaining syntax `@(Foo->'%(Bar).Baz'->Replace('Baz', 'FooBar'))` and creates an ItemExpressionCapture instance for each function call, then a higher-level ItemExpressionCapture instance whose Captures property lists all of them. Expander\<P, I>.ItemExpander.ExpandExpressionCapture\<S> calls the functions.  <https://github.com/dotnet/msbuild/blob/9802d43bc30b997b75c5efa310a9d379bd56162f/src/Build/Evaluation/ExpressionShredder.cs#L172-L173> <https://github.com/dotnet/msbuild/blob/9802d43bc30b997b75c5efa310a9d379bd56162f/src/Build/Evaluation/ExpressionShredder.cs#L193> <https://github.com/dotnet/msbuild/blob/9802d43bc30b997b75c5efa310a9d379bd56162f/src/Build/Evaluation/ExpressionShredder.cs#L260-L263> <https://github.com/dotnet/msbuild/blob/9802d43bc30b997b75c5efa310a9d379bd56162f/src/Build/Evaluation/Expander.cs#L1877-L1880>

Expander\<P, I>.ItemExpander.IntrinsicItemFunctions\<S>.GetItemTransformFunction searches for the function in IntrinsicItemFunctions\<S>, or if not found, uses ExecuteStringFunction, which then searches for the function in System.String.  <https://github.com/dotnet/msbuild/blob/9802d43bc30b997b75c5efa310a9d379bd56162f/src/Build/Evaluation/Expander.cs#L2076-L2087>

GetItemTransformFunction also recognizes the `@(MyItem->MetadataName())` function-call syntax for reading well-known item metadata such as `Filename`.  I guess that is an undocumented legacy feature; unlike `@(MyItem->Metadata("MetadataName"))` and `@(MyItem->'%(MetadataName)')`, it does not support `RecursiveDir` or custom metadata.  <https://github.com/dotnet/msbuild/blob/9802d43bc30b997b75c5efa310a9d379bd56162f/src/Build/Evaluation/Expander.cs#L2069-L2073>

**Reply by ltrzesniewski** [reply]
Thanks for the analysis.

> Expander\<P, I\>.ItemExpander.IntrinsicItemFunctions\<S\>.GetItemTransformFunction searches for the function in IntrinsicItemFunctions\<S\>, or if not found, uses ExecuteStringFunction, which then searches for the function in System.String.

That's good to know, so MSBuild always tries to call item functions first, then falls back to string functions.

> It doesn't look like a coincidence to me, but not really documented either, so I'm not sure it will work in future versions.

That's actually why I asked the question: since it's not documented I don't know if I can rely on it in future versions, or if it's just a side effect of the current implementation. Ideally, I'd like someone from MS to comment on it (and maybe document it if it's an intended feature).
